### PR TITLE
Turn on clang-tidy in the CI build

### PR DESCRIPTION
**Pull request by TrentHouliston**
This will enforce all the clang-tidy rules we have (so this won't merge yet)
Keep merging fixes into master until this PR has a green tick!

### CODE REVIEW

**Comment by KipHamiltons** [comment]
The size of this is absolutely insane. I reckon we should break this into multiple PRs, only enabling the CI build at the last one.
This one is already huge enough...

**Comment by KipHamiltons** [comment]
Also, note to any reviewer - it's probably easiest to review this commit by commit.
The nested namespace change makes things very messy, so if it were me, I would just take that commit on faith because I can certify that after that commit, it built, and I didn't change anything except namespaces and the end namespace comments.

**Comment by KipHamiltons** [comment]
Oh shit. I just understood what the intention of this PR was....
well, now that I have effectively hijacked it, I suppose we should turn this PR into one of the ones which merges changes :sweat_smile:

**Comment by TrentHouliston** [comment]
Nothing a force push can't solve!

**Comment by tayloryoung6396** [comment]
@TrentHouliston @KipHamiltons What is the state of this PR? As I understand it's no longer needed

**Comment by KipHamiltons** [comment]
@tayloryoung6396 It is needed, but it has to be updated. What this PR is going to do is add clang-tidy to the CI checks. However, we have to fix the clang-tidy warnings which are in the codebase atm for that to be useful. So this is waiting for #734 (and the ones which come after it) to fix those warnings. Then, we'll update this PR so that it adds clang-tidy to the github actions workflows, instead of buildkite.
Long story short, keep this PR, and approve my clang-tidy fix PRs as they come in :sweat_smile:

**Comment by CMurtagh-LGTM** [comment]
Is this already a feature?
https://github.com/NUbots/NUbots/blob/main/.github/workflows/nubots.yaml#L145
